back-end/utils/top-layer/UserUtil.js

```diff
@@ -432,61 +432,55 @@ const getLengthUserTransactionsHistoryForRedisM5RU = (email, filters) => {
 
 const sellShareEvent = async (
   user,
-  totalPendingPrice,
+  totalCashChange,
   companyCode,
-  quantity,
-  recentPrice
+  quantity
 ) => {
   try {
-    user.cash += totalPendingPrice;
+    // Update user cash
+    await prisma.user.update({
+      where: {
+        id: user.id,
+      },
+      data: {
+        cash: user.cash + totalCashChange,
+      }
+    });
+
+    // Find the stock user want to sell and update its properties (quantity and average price)
     const sellShare = user.shares.find(
       (share) => share.companyCode === companyCode
     );
-    const sellShareIndex = user.shares.indexOf(sellShare);
 
     if (!sellShare) {
       return new Error("Couldn't find share");
     } else if (quantity > sellShare.quantity) {
       return new Error("Not enough available shares");
+    } else if (user.cash + totalCashChange < 0) {
+      return new Error("Something is wrong with user's cash")
     }
 
     if (quantity === sellShare.quantity) {
-      await prisma.share.delete({
-        where: {
-          companyCode: companyCode
-        }
-      });
-
-      user.shares.slice(sellShareIndex, 1);
-      await prisma.user.update({
-        where: {
-          id: user.id
-        },
-        data: {
-          shares: user.shares
-        }
-      });
-    } else {
-      sellShare.quantity -= quantity;
-      await prisma.share.update({
-        where: {
-          id: sellShare.id
-        },
-        data: {
-          quantity: sellShare.quantity
-        }
-      });
-
-      user.shares[sellShareIndex] = sellShare;
-      await prisma.user.upate({
-        where: {
-          id: user.id
-        },
-        data: {
-          shares: user.shares
-        }
-      });
+      user.shares = user.shares.filter((share) => share !== sellShare);
```

Comment:
This new flow is wrong because it does not delete the share if the quantity is equal to zero. Revert to the old flow and just delete the user update parts.